### ISSUE TITLE
Constraining drag not working correctly?

### ISSUE THREAD

**Issue by bdorninger** [chrispahm/chartjs-plugin-dragdata#90]
**Describe the bug**

Drag constrained by "onDrag" does not work as intended.
If dragging a point, the constraint works as expected DURING drag (point keeps glued to upper/lower limit), but on dragEnd - if the cursor is beyond a limit - the point suddenly jumps to a "forbidden" value

**To Reproduce**
I have the code in a public [github](https://github.com/bdorninger/chartjs-eval/tree/master)
Also on [stackblitz](https://stackblitz.com/edit/chartjs-eval?file=index.ts)

The constraint of the drag is done via (index.ts:403) :

```
onDrag: function (e, datasetIndex, index, value: number | Point) {
      e.target.style.cursor = 'grabbing';
          if (typeof value === 'number') {
            return value >= 2 && value <= 19;
          }
          return value.y >= 2 && value.y <= 19;
        },

```

**Expected behavior**

If beyond a limit point during drag, the point shall stay there, regardless if the cursor moved further beyond on dragEnd.

**Comment by chrispahm** [comment]
Hey @bdorninger,

That is indeed strange, I tried to reproduce with a simple example, but unfortunately couldn’t:
https://jsfiddle.net/92vwpr31/2/

Maybe it's the combination of the zoom/crosshair/drag-data pluging that is causing this, but I'm just guessing for now

**Comment by bdorninger** [comment]
Hi chris

It's not the other plugins, because the error persists when removing the other plugins.

After further looking into it , I heavily suspect the usage of object data being the cause. Your fiddle uses number arrays for Y and fixed categories for the X axis.
My sample works with point data objects {x: number, y: number} on having linear axes for both x and y (https://www.chartjs.org/docs/latest/general/data-structures.html#object)

Forked your fiddle to : https://jsfiddle.net/w8fj2hm1/ , modified it and could reproduce the issue

I suppose, plugin code on releasing the drag assumes data to be numbers only. 




> Hey @bdorninger,
> 
> That is indeed strange, I tried to reproduce with a simple example, but unfortunately couldn’t: https://jsfiddle.net/92vwpr31/2/
> 
> Maybe it's the combination of the zoom/crosshair/drag-data pluging that is causing this, but I'm just guessing for now

**Comment by chrispahm** [comment]
Thanks @bdorninger for digging into this! The "object theory" would somewhat make sense, could be that we need to make a deep copy of the object to make sure it's not modified afterwards... I'll have a look

**Comment by chrispahm** [comment]
Alright, I now made a minor change creating a deep copy of the original data object. That solves the bug on my end, would be nice if you could confirm!
[chartjs-plugin_dragdata-v2.2.5_beta.zip](https://github.com/chrispahm/chartjs-plugin-dragdata/files/7994814/chartjs-plugin_dragdata-v2.2.5_beta.zip)

Best
Chris

**Comment by bdorninger** [comment]
Looks good!
I had done a similar change in the meantime, which also solved it:
```
const dp = chartInstance.data.datasets[datasetIndex].data[index];
const dataPoint = typeof dp === 'object' ? { ...dp } : dp;
```
Thx for your quick help!

regards,
Bernhard

**Comment by chrispahm** [comment]
Yes that's great, I just did a quick `JSON.stringify(JSON.parse())`

**Comment by bdorninger** [comment]
..yours is a real deep copy, mine is just flat ;-)

**Comment by chrispahm** [comment]
The flat copy may have the benefit of not stripping any getters/setters from the original object. Not sure if it's a common pattern to use them in this context but I guess it could be annoying if the plug-in would unintentionally remove those

**Comment by ggomaeng** [comment]
@chrispahm Is this fix available on npm? Facing the same issue

**Comment by ggomaeng** [comment]
update: forked the repo and fixed it myself, but it would be nice if it's published on npm 👍

**Comment by artus9033** [comment]
Hi @ggomaeng, I took a look at this issue and you identified the culprit correctly - it is the conditional code that operates only on object-style data points and does so before the callback is actually executed. Since objects are complex values and the code operates on the original objects instead of copies, the effects are effective even though they not always should be.

It turns out that there was a different issue #96 that should solve this by copying the data point inside `calcPosition`, which effectively should resolve this issue as well. I don't think that the point inside `updateData` also needs to be cloned (for performance reasons), since `roundValue`, `calcRadar` &`calcPosition` should not mutate their parameters.

In the next few days, I am planning to update & merge this and a few other PRs, and I will come back to this issue afterwards - I believe at the moment that this will be resolved as #96 gets merged.

**Comment by artus9033** [comment]
I tested this problem on the current state of the master branch (8a62d9f9f651edf6b8edfcb3bf4e1e9c19d7cea4) and cannot reproduce this problem - as expected, now it seems to work correctly, so I'm closing the issue.

![mre-post](https://github.com/user-attachments/assets/ae8aa9bf-83c9-42d4-89d1-2752ca0fa1da)
